### ISSUE TITLE
Unable to save project

### ISSUE THREAD

**Issue by BeatBurgener** [brandoncomputer/powershell-designer#26]
Hi there

I just stumbled up on this project and tried to give it a go. But I could not get to the point to save the form - which is the basis for everything else ...

Maybe related to country specific settings?

20241029 08:21:26: Cannot convert value "800; 600" to type "System.Int32". Error: "Input string was not in a correct format." At line:4191 char:25 + $n[0] = [math]::round(($n[0]/1) / $ctscale) + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  + CategoryInfo : InvalidArgument: (:) [], RuntimeException  + FullyQualifiedErrorId : InvalidCastFromStringToInteger  

It looks like the main form size cannot be converted. 

Any idea?
Thank you
Beat

**Comment by brandoncomputer** [comment]
Need to fix localization, in this case a semicolon is used due to region rather than a comma. Need to acquire this as a variable and use that instead.

**Comment by brandoncomputer** [comment]
I believe we are looking for [cultureinfo]::CurrentCulture.TextInfo.ListSeparator

**Comment by brandoncomputer** [comment]
@BeatBurgener can you run the following in powershell and confirm it returns semicolon?

[cultureinfo]::CurrentCulture.TextInfo.ListSeparator

**Comment by brandoncomputer** [comment]
Appreciate users posting country localization codes eg. en-US, en-GB along with the output of `[cultureinfo]::CurrentCulture.TextInfo.ListSeparator`

**Comment by brandoncomputer** [comment]
I believe this is fixed with 2.7.0. Leaving open for now.

**Comment by brandoncomputer** [comment]
Not fixed. Need to correct in functions.psm1. Need to make certain global is declared before module load or reference. Need to update functions in module. 2.7.1 should fix.

**Comment by BeatBurgener** [comment]
yes, you are right, it is a ";"

PS C:\Windows\System32\WindowsPowerShell\v1.0> [cultureinfo]::CurrentCulture.TextInfo.ListSeparator
;
PS C:\Windows\System32\WindowsPowerShell\v1.0>

**Comment by brandoncomputer** [comment]
@BeatBurgener 
This is probably fixed. Kindly run 
`Install-Module powershell-designer -force` 
`powershell-designer`
And test it out.

**Comment by BeatBurgener** [comment]
Unfortunately not:
20241115 18:26:31: Cannot convert value "800; 600" to type "System.Int32". Error: "Input string was not in a correct format." At line:4191 char:25 + $n[0] = [math]::round(($n[0]/1) / $ctscale) + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  + CategoryInfo : InvalidArgument: (:) [], RuntimeException  + FullyQualifiedErrorId : InvalidCastFromStringToInteger  

vs before:

20241029 08:22:11: Cannot convert value "1000; 600" to type "System.Int32". Error: "Input string was not in a correct format." At line:4191 char:25 + $n[0] = [math]::round(($n[0]/1) / $ctscale) + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  + CategoryInfo : InvalidArgument: (:) [], RuntimeException  + FullyQualifiedErrorId : InvalidCastFromStringToInteger  

Looks like the split does not work ... 

If I may add: writing errors to windows\system32\ or \Windows\System32\WindowsPowerShell\v1.0 is a suboptimal choice, as this would require powershell-designer to be run as administrator, which is bad practice ...

Thank you
Beat

**Comment by brandoncomputer** [comment]
@BeatBurgener 

Hi Beat,

Thanks so much for the feedback concerning the log file. I will open an issue to address that.

For the issue at hand, try the following:

Close Powershell, open powershell-designer, try again. Sometimes it won't register the new module version until a new session is started.

In all seriousness, I really do appeciate the feedback. 

Thanks,
Brandon

**Comment by BeatBurgener** [comment]
Brandon, 
I did verify with TaskManager that there is no powershell.exe left, still, I get the same error. Any option to enable debugging in your application to log what is happening where?

Thank you
Beat

**Comment by brandoncomputer** [comment]
This is again in a 'probably fixed' status. Just published. 

Updating the repository now.

**Comment by BeatBurgener** [comment]
Thank you Brandon for your efforts, still no luck ... same error.

**Comment by brandoncomputer** [comment]
Hi Beat,

This problem is larger than I first expected, the trouble is the list seperator is actually also stored in static string data, like such

`ConvertFrom-WinFormsXML -Reference refs -Suppress -Xml @"
<Form Name="MainForm" MinimumSize="800,600" Size="1526,939" Tag="IsMDIContainer, DPIAware, VisualStyle" WindowState="Maximized" Text="PowerShell Designer" IsMDIContainer="True">`

All of these commas would have to be converted to semicolons some how, for a hope for the software to work, and even then, if you were to distribute your outputted script to the United States, it would not run on computers in that localization. 

I need to consider the way forward here, it's not acceptable that the software simply does not work for so much of the globe, but I'm also not certain the XML format will survive the potential fix, and the whole project is wrapped around XML. 

This is my first encounter with this trouble, let me see if I can find how others have mitigated the trouble in the past, I can't be the first person with the problem.

**Comment by brandoncomputer** [comment]
The trouble continues... even after converting static data, items like system.drawing.point only allow comma's.

20241116 18:08:35: ForEach-Object : Exception lors de la définition de «MinimumSize»: «Impossible de convertir la valeur «800;600» en type  «System.Drawing.Size». Erreur: «800;600 n'est pas une valeur valide pour Int32.»» Au caractère Ligne:431 : 21 + $Xml.Attributes

**Comment by BeatBurgener** [comment]
Well, I'll do some dev too, although I have no education in this area... a few ideas, but not sure how this value "800;600" is passed / received by your application. If this is a single value for the main forms width;height, then this is a problem. Preferable this would be two distinct properties of the form, where you would assign a dedicated variable instead of parsing it. Or the dimension of the form could be received as a array or better an object with properties instead of a string. PS/.Net is about objects anyway, although it looks like based on this .Net Doc that it is indeed one value for "Size" e.g width/height.: 

https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.form.size?view=windowsdesktop-8.0

How unfortunate is that? Especially, if that function is country code aware ... for what reason?
E.g. one option could be to use regex to extract what ever separator is used to separate width / height ... without looking at your code, maybe this is a bad advice... e.g. just a hint.

**Comment by brandoncomputer** [comment]
This will probably be fixed in a few minutes. The issue with exceptions.txt will return.

**Comment by brandoncomputer** [comment]
> Well, I'll do some dev too, although I have no education in this area... a few ideas, but not sure how this value "800;600" is passed / received by your application. If this is a single value for the main forms width;height, then this is a problem. Preferable this would be two distinct properties of the form, where you would assign a dedicated variable instead of parsing it. Or the dimension of the form could be received as a array or better an object with properties instead of a string. PS/.Net is about objects anyway, although it looks like based on this .Net Doc that it is indeed one value for "Size" e.g width/height.:
> 
> https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.form.size?view=windowsdesktop-8.0
> 
> How unfortunate is that? Especially, if that function is country code aware ... for what reason? E.g. one option could be to use regex to extract what ever separator is used to separate width / height ... without looking at your code, maybe this is a bad advice... e.g. just a hint.

I'm working on this now, fortunately many of my assumptions were false. This is simply a localization error with the PropertyGrid control.

**Comment by brandoncomputer** [comment]
@BeatBurgener
This is probably fixed. Kindly run
Install-Module powershell-designer -force
Close powershell.
Launch powershell.
powershell-designer
And test it out.

**Comment by BeatBurgener** [comment]
I'm sorry, but no change:
20241117 20:42:45: Cannot convert value "800; 600" to type "System.Int32". Error: "Input string was not in a correct format." At line:4191 char:25 + $n[0] = [math]::round(($n[0]/1) / $ctscale) + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  + CategoryInfo : InvalidArgument: (:) [], RuntimeException  + FullyQualifiedErrorId : InvalidCastFromStringToInteger

**Comment by brandoncomputer** [comment]
Yeah, not fixed. Publishing error, still version 2.7.2 instead of 2.7.3, you can manually patch from the repo here or wait for update.

**Comment by brandoncomputer** [comment]
For some reason publishing fails when my language mode is set to French, set back to English and it published.

Should be good to try again

**Comment by BeatBurgener** [comment]
@brandoncomputer - sorry to inform you: nope ....

**Comment by brandoncomputer** [comment]
@BeatBurgener Can you verify tthe module version is 2.7.3?

**Comment by brandoncomputer** [comment]
> @BeatBurgener Can you verify the module version is 2.7.3 (I'm pretty sure this is a publishing error)?

Nope: I think I was fooled. [cultureinfo]::CurrentCulture.TextInfo.ListSeparator is returning a comma ',' right now when it should be returning a semicolon. Let me work on this.

**Comment by BeatBurgener** [comment]
Version    Name                                Repository           Description
-------    ----                                ----------           -----------
2.7.3      powershell-designer                 PSGallery            The best free form designer for powershell.

**Comment by brandoncomputer** [comment]
Nope: I think I was fooled. [cultureinfo]::CurrentCulture.TextInfo.ListSeparator is returning a comma ',' right now when it should be returning a semicolon. Let me work on this.

**Comment by BeatBurgener** [comment]
Sorry that this gives you such a hard time .. as a workaround, one might override the culture while starting the PS. Ugly workaround, I agree, but would make it run for now ... maybe ... potentially ... Culture can be overridden by thread ...

**Comment by brandoncomputer** [comment]
Changed region, and it is still working, [cultureinfo]::CurrentCulture.TextInfo.ListSeparator is returning a semicolon... can you open designer.ps1 in an editor and look at line 4228

Should be 

`$n = ($node.Size).replace(';',',').split(',')`

**Comment by brandoncomputer** [comment]
And I see the problem. How did that happen?

**Comment by brandoncomputer** [comment]
> Sorry that this gives you such a hard time .. as a workaround, one might override the culture while starting the PS. Ugly workaround, I agree, but would make it run for now ... maybe ... potentially ... Culture can be overridden by thread ...

I won't suggest that, I need to fix it :) However, if that works for you great, it'll just be tougher for me to verify it's fixed ;p

**Comment by BeatBurgener** [comment]
... I'll let you figure it out ... will not tinker with Culture then  ;-)

**Comment by brandoncomputer** [comment]
Changed region, and it is still working, [cultureinfo]::CurrentCulture.TextInfo.ListSeparator is returning a semicolon... can you open designer.ps1 in an editor and look at line 4228

Should be

$n = ($node.Size).replace(';',',').split(',')

**Comment by brandoncomputer** [comment]
@BeatBurgener 
I'm not able to replicate the trouble at this point, what are your Windows Region settings? The software is working properly for United States (Comma Seperator) and France (Semicolon Seperator)

**Comment by BeatBurgener** [comment]
PS C:\Windows\System32\WindowsPowerShell\v1.0> Get-Culture

LCID             Name             DisplayName
----             ----             -----------
2055             de-CH            German (Switzerland)

You can use set-culture to test with "my" environment

**Comment by brandoncomputer** [comment]
Can you confirm in Exceptions.txt you are getting an error similar to this?

20241119 18:20:37: ForEach-Object : Ausnahme beim Festlegen von "ShortcutKeys": "Der Wert "Ctrl+D" kann nicht in den Typ  "System.Windows.Forms.Keys" konvertiert werden. Fehler: "Der angeforderte Wert "Ctrl" konnte nicht gefunden werden."" In C:\Users\brand\OneDrive\Documents\WindowsPowerShell\Modules\powershell-designer\2.7.3\functions\functions.psm1:2063  Zeichen:21 + $Xml.Attributes | ForEach-Object { + ~~~~~~~~~~~~~~~~  + CategoryInfo : NotSpecified: (:) [ForEach-Object], SetValueInvocationException  + FullyQualifiedErrorId : ExceptionWhenSetting,Microsoft.PowerShell.Commands.ForEachObjectCommand 

This has to do with German using Strg instead of Ctrl as a key reference.

**Comment by BeatBurgener** [comment]
Nope, didn't see that error (would also create a popup I assume, not silently written to the log)

**Comment by brandoncomputer** [comment]
Hi @BeatBurgener 

I've found that PowerShell 7x has cleared up much of these regional difficulties, I even reverted the changes back to 2.6.9 and the software seems to be working.

Can you try testing in PowerShell 7x?

Thanks,
Brandon

**Comment by brandoncomputer** [comment]
My test environment is borked. I can't get it to behave properly.

I do suggest trying PowerShell 7.5 and up, let me know the results. I'm trying to fix my environment, something is broken, or rather, I can't get it to break the same as before.

**Comment by BeatBurgener** [comment]
7.5 is in preview only, not really sure i'd like to go this route ... got 7.4.6. installed for now ... same error

**Comment by brandoncomputer** [comment]
I just stood up a second test environment, changed all the settings to the best of my ability to German/Switzerland, installed the module, and did not experience the issue

Do you have a second computer to try the module on?

**Comment by brandoncomputer** [comment]
@BeatBurgener 

Minor fix published today, not certain if it's relevant to what you're experiencing. Might work.
